src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionLiteral.cs

```diff
@@ -0,0 +1,121 @@
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal sealed partial class LocalRewriter
+    {
+        public override BoundNode? VisitArrayOrSpanCollectionLiteralExpression(BoundArrayOrSpanCollectionLiteralExpression node)
+        {
+            Debug.Assert(!_inExpressionLambda);
+            Debug.Assert(node.Type is { });
+
+            var syntax = node.Syntax;
+            var collectionType = node.Type;
+            var spanConstructor = node.SpanConstructor;
+
+            var arrayType = collectionType as ArrayTypeSymbol;
+            if (arrayType is null)
+            {
+                Debug.Assert(spanConstructor is { });
+                Debug.Assert(TypeSymbol.Equals(collectionType, spanConstructor.ContainingType, TypeCompareKind.AllIgnoreOptions));
+
+                var elementType = ((NamedTypeSymbol)collectionType).TypeArgumentsWithAnnotationsNoUseSiteDiagnostics[0];
+                arrayType = ArrayTypeSymbol.CreateSZArray(_compilation.Assembly, elementType);
+            }
+
+            var initializers = node.Initializers;
+            int arrayLength = initializers.Length;
+            // PROTOTYPE: Should [] be emitted as Array.Empty<T>()?
+            var initialization = (arrayLength == 0)
+                ? null
+                : new BoundArrayInitialization(
+                    syntax,
+                    isInferred: false,
+                    initializers.SelectAsArray(e => VisitExpression(e)));
+            var array = new BoundArrayCreation(
+                syntax,
+                ImmutableArray.Create<BoundExpression>(
+                    new BoundLiteral(
+                        syntax,
+                        ConstantValue.Create(arrayLength),
+                        _compilation.GetSpecialType(SpecialType.System_Int32))),
+                initialization,
+                arrayType)
+            { WasCompilerGenerated = true };
+
+            if (spanConstructor is null)
+            {
+                return array;
+            }
+
+            Debug.Assert(TypeSymbol.Equals(array.Type, spanConstructor.Parameters[0].Type, TypeCompareKind.AllIgnoreOptions));
+            return new BoundObjectCreationExpression(syntax, spanConstructor, array);
+        }
+
+        public override BoundNode? VisitCollectionInitializerCollectionLiteralExpression(BoundCollectionInitializerCollectionLiteralExpression node)
+        {
+            Debug.Assert(!_inExpressionLambda);
+            Debug.Assert(node.Type is { });
+
+            var syntax = node.Syntax;
+            var collectionType = node.Type;
+
+            BoundExpression rewrittenReceiver;
+            if (collectionType is TypeParameterSymbol typeParameter)
+            {
+                // PROTOTYPE: If we support _inExpressionLambda, see VisitNewT()
+                // which does not call MakeNewT() in that case.
+                rewrittenReceiver = MakeNewT(syntax, typeParameter);
```

Comment:
I didn't understand why we need a separate path for type parameters here instead of being able to use a CollectionCreation node in all cases.

Reply:
Updated to use `CollectionCreation` in all cases, thanks.